docs/proposal-007-apply.md

```diff
@@ -0,0 +1,173 @@
+once, we need to instead gradually support properties and resources of a
+cluster.
+
+This leads to the question of whether the current `ClusterConfig` is a good
+structure for a config meant for reconciliation.
+
+## Proposal
+
+Introduce a subcommand `apply` which has only a config file argument.
+`apply` would reconcile desired with real-world state
+[as described in the motivation](#motivation).
+The exact details per resource will potentially differ depending on
+the resources and won't be enumerated here.
+
+1. `apply` is put behind the explicit "experimental" flag (cf `enable profile/repo`)
+2. Take full ownership of `eksctl`-managed resources (i.e. delete missing resources by default)
+3. Options for non-`eksctl`-managed resources as well as resources where
+   ownership is unknowable are covered below
+4. Changing immutable fields through recreation is the default
+5. We gradually expand `apply` support to different parts of the cluster
+6. As we expand `apply` support, we reevaluate and update the config structure as
+   necessary in a new API version.
+
+Some further discussion/justification follows.
+
+### Experimental flag
+
+We make it explicit that `apply` is experimental, very open to feedback and
+subject to change.
+
+### Ownership
+
+Ownership comes into play most importantly with deletion.
+
+#### `eksctl`-managed resources
+
+Given the config:
+
+```
+nodeGroups:
+  - name: ng-2
+    desiredCapacity: 1
+```
+
+we discover an existing nodegroup `ng-1` with the `eksctl.io/owned: true` tag (for example),
+do we delete it?
+
+The proposed answer is yes, it should be deleted. With `apply` we assume that all
+previous changes initiated by `eksctl` were initiated through the same config
+file/reconciliation process and that this nodegroup must have been
+deleted from this file with the intention of deleting it from the cluster.
+
+##### Non-authoritative resources
+
+There is precedent for adding the ability to specify resources in a
+non-authoritative list. For example, the [`google` terraform provider supports IAM with 3 different levels of authoritativeness](
+https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/storage_bucket_iam#google_storage_bucket_iam_policy#iam-policy-for-cloud-storage-bucket).
+
+Initial `apply` support will focus on authoritative resource specification.
+
+#### Non-taggable resources
+
+Some resources may not be taggable as owned by `eksctl`.
+If there weren't a `tags` field on addons, for example. `eksctl`-created
+addons would then be indistinguishable from addons created in the AWS console.
+There are at least 2 ways to deal with this case:
+
+##### `eksctl` owns everything
+
+With this option, we assume that `eksctl` owns the set of such resources and
+delete any not listed in the config.
+
+##### `terraform`-style state
+
+Storing some kind of state/config "somewhere" is another option and has been discussed in
+https://github.com/weaveworks/eksctl/issues/642
```

Comment:
I like the idea (from that linked thread) of representing the full cluster in a ConfigMap in the cluster itself (with a view to use a CRD in the future) and reconciling config changes against that diff. Would be good if we could get more information around the difficulties found in the original exploration and whether this needs to be a "thing" in and of itself:

> I'm removing this from the milestone, as MVP is not a simple as I hoped for. We should also create a design document for this actually.